auth/service.go

```diff
@@ -1030,3 +1035,132 @@ func DecodeDomainUserID(domainUserID string) (string, string) {
+	switch entityType {
+	case ThingType:
+		// Remove policy of groups
+		req := PolicyReq{
+			SubjectType: GroupType,
+			Object:      id,
+			ObjectType:  ThingType,
+		}
+
+		if err := svc.DeletePolicy(ctx, req); err != nil {
+			return err
+		}
+
+		// Remove policy from domain
+		req.SubjectType = DomainType
+		if err := svc.DeletePolicy(ctx, req); err != nil {
+			return err
+		}
+
+		// Remove policy of users
+		req.SubjectType = UserType
+		if err := svc.DeletePolicy(ctx, req); err != nil {
+			return err
+		}
+
+		return nil
+	case UserType:
+		domainsPage, err := svc.domains.ListDomains(ctx, Page{SubjectID: id, Limit: defLimit})
+		if err != nil {
+			return err
+		}
+
+		if domainsPage.Total > defLimit {
+			for i := defLimit; i < int(domainsPage.Total); i += defLimit {
+				page := Page{SubjectID: id, Offset: uint64(i), Limit: defLimit}
+				dp, err := svc.domains.ListDomains(ctx, page)
+				if err != nil {
+					return err
+				}
+				domainsPage.Domains = append(domainsPage.Domains, dp.Domains...)
+			}
+		}
+
+		for _, domain := range domainsPage.Domains {
+			policy := PolicyReq{
+				Subject:     EncodeDomainUserID(domain.ID, id),
+				SubjectType: UserType,
+				ObjectType:  ThingType,
+			}
+			if err := svc.agent.DeletePolicy(ctx, policy); err != nil {
+				return err
+			}
+
+			policy.ObjectType = GroupType
+			if err := svc.agent.DeletePolicy(ctx, policy); err != nil {
+				return err
+			}
+
+			policy.Object = domain.ID
+			policy.ObjectType = DomainType
+			if err := svc.agent.DeletePolicy(ctx, policy); err != nil {
+				return err
+			}
+		}
+
+		req := PolicyReq{
+			Subject:     id,
+			SubjectType: UserType,
+			ObjectType:  ThingType,
+		}
+		if err := svc.agent.DeletePolicy(ctx, req); err != nil {
+			return err
+		}
+		req.ObjectType = GroupType
+		if err := svc.agent.DeletePolicy(ctx, req); err != nil {
+			return err
+		}
+		req.ObjectType = DomainType
+		if err := svc.agent.DeletePolicy(ctx, req); err != nil {
+			return err
+		}
+		req.ObjectType = PlatformType
+		req.Object = MagistralaObject
+		if err := svc.agent.DeletePolicy(ctx, req); err != nil {
+			return err
+		}
+
+		if err := svc.domains.DeleteUserPolicies(ctx, id); err != nil {
+			return err
+		}
+
+		return nil
+	case GroupType:
+		// Remove policy of child groups
+		req := PolicyReq{
+			SubjectType: GroupType,
+			Subject:     id,
+			ObjectType:  GroupType,
+		}
+		if err := svc.DeletePolicy(ctx, req); err != nil {
+			return err
+		}
+
+		// Remove policy of things
+		req.ObjectType = ThingType
+		if err := svc.DeletePolicy(ctx, req); err != nil {
+			return err
+		}
+
+		// Remove policy from domain
+		req.SubjectType = DomainType
+		req.ObjectType = GroupType
+		if err := svc.DeletePolicy(ctx, req); err != nil {
+			return err
+		}
+
+		// Remove policy of users
+		req.SubjectType = UserType
+		if err := svc.DeletePolicy(ctx, req); err != nil {
+			return err
+		}
```

Comment:
```suggestion
	case GroupType:
		// Remove policies where group is subject and object is another child group.
		req := PolicyReq{
			SubjectType: GroupType,
			Subject:     id,
			ObjectType:  GroupType,
		}
		if err := svc.DeletePolicy(ctx, req); err != nil {
			return err
		}

		// Remove policies where group is subject and object is thing.
		req.ObjectType = ThingType
		if err := svc.DeletePolicy(ctx, req); err != nil {
			return err
		}

		// Remove all policies where group is object.
		req = PolicyReq{
			Object:     id,
			ObjectType: GroupType,
		}
		if err := svc.DeletePolicy(ctx, req); err != nil {
			return err
		}
```

The above suggestin will work , if we have changes in svc.agent.DeletePolicy show below 

```
func (pa *policyAgent) DeletePolicy(ctx context.Context, pr auth.PolicyReq) error {
	req := &v1.DeleteRelationshipsRequest{
		RelationshipFilter: &v1.RelationshipFilter{
			ResourceType:       pr.ObjectType,
			OptionalResourceId: pr.Object,
		},
	}

	if pr.Relation != "" {
		req.RelationshipFilter.OptionalRelation = pr.Relation
	}

	if pr.SubjectType != "" {
		req.RelationshipFilter.OptionalSubjectFilter = &v1.SubjectFilter{
			SubjectType: pr.SubjectType,
		}
		if pr.Subject != "" {
			req.RelationshipFilter.OptionalSubjectFilter.OptionalSubjectId = pr.Subject
		}
		if pr.SubjectRelation != "" {
			req.RelationshipFilter.OptionalSubjectFilter.OptionalRelation = &v1.SubjectFilter_RelationFilter{
				Relation: pr.SubjectRelation,
			}
		}
	}

	if _, err := pa.permissionClient.DeleteRelationships(ctx, req); err != nil {
		return errors.Wrap(errRemovePolicies, handleSpicedbError(err))
	}
	return nil
}
```

Reply:
This automatically deletes the other group entities